lib/pure/optionals.nim

```diff
@@ -0,0 +1,262 @@
+  maybe.val
+
+proc `[]`*[T](maybe: Maybe[T]): T =
+  ## Returns the value of `maybe`. Raises ``FieldError`` if it is `nothing`.
+  if not maybe:
+    raise newException(FieldError, "Can't obtain a value from a `nothing`")
+  maybe.val
+
+
+template `or`*[T](maybe: Maybe[T], default: T): T =
+  ## Returns the value of `maybe`, or `default` if it is `nothing`.
+  if maybe: maybe.val
+  else: default
+
+template `or`*[T](a, b: Maybe[T]): Maybe[T] =
+  ## Returns `a` if it is `just`, otherwise `b`.
+  if a: a
+  else: b
+
+template `?=`*(into: expr, maybe: Maybe): bool =
+  ## Returns ``true`` if `maybe` isn't `nothing`.
+  ##
+  ## Injects a variable with the name specified by the argument `into`
+  ## with the value of `maybe`, or its type's default value if it is `nothing`.
+  ##
+  ## .. code-block:: nim
+  ##
+  ##   proc message(): ?string =
+  ##     just "Hello"
+  ##
+  ##   if m ?= message():
+  ##     echo m
+  var into {.inject.}: type(maybe.val)
+  if maybe:
+    into = maybe.val
+  maybe
+
+
+proc `==`*(a, b: Maybe): bool =
+  ## Returns ``true`` if both ``Maybe`` are `nothing`,
+  ## or if they have equal values
+  (a.has and b.has and a.val == b.val) or (not a.has and not b.has)
+
+proc `$`[T](maybe: Maybe[T]): string =
+  ## Converts to string: `"just(value)"` or `"nothing(type)"`
+  if maybe.has:
+    "just(" & $maybe.val & ")"
+  else:
+    "nothing(" & T.name & ")"
+
+
+when isMainModule:
+  template expect(E: expr, body: stmt) =
+    try:
+      body
+      assert false, E.type.name & " not raised"
+    except E:
+      discard
+
+
+  block: # example
+    proc find(haystack: string, needle: char): ?int =
+      for i, c in haystack:
+        if c == needle:
+          return just i
+
+    assert("abc".find('c')[] == 2)
+
+    let result = "team".find('i')
+
+    assert result == nothing(int)
+    assert result.has == false
+
+    if pos ?= "nim".find('i'):
+      assert pos is int
+      assert pos == 1
+    else:
+      assert false
+
+    assert(("team".find('i') or -1) == -1)
+    assert(("nim".find('i') or -1) == 1)
+
+  block: # just
+    assert just(6)[] == 6
+    assert just("a").unsafeVal == "a"
+    assert just(6).has
+    assert just("a")
+
+  block: # nothing
+    expect FieldError:
+      discard nothing(int)[]
+    assert(not nothing(int).has)
+    assert(not nothing(string))
+
+  block: # equality
+    assert just("a") == just("a")
+    assert just(7) != just(6)
+    assert just("a") != nothing(string)
+    assert nothing(int) == nothing(int)
+
+    when compiles(just("a") == just(5)):
+      assert false
+    when compiles(nothing(string) == nothing(int)):
+      assert false
+
+  block: # stringification
+    assert "just(7)" == $just(7)
+    assert "nothing(int)" == $nothing(int)
+
+  block: # or
+    assert just(1) or just(2) == just(1)
+    assert nothing(string) or just("a") == just("a")
+    assert nothing(int) or nothing(int) == nothing(int)
+    assert just(5) or 2 == 2
```

Comment:
Hm, this is weird. Traditionally, if we already have a value, injecting a default should not have any effect. I would expect `just(5) or 2 == 5`.